// discord-gamecord/src/GuessThePokemon.js
const { EmbedBuilder, AttachmentBuilder } = require('discord.js');
const fetch = require('node-fetch');
const events = require('events');

module.exports = class GuessThePokemon extends events {
  constructor(options = {}) {
    if (!options.isSlashGame) options.isSlashGame = false;
    if (!options.message) throw new TypeError('NO_MESSAGE: No se proporcionó la opción de mensaje.');
    if (typeof options.message !== 'object') throw new TypeError('INVALID_MESSAGE: La opción de mensaje debe ser un objeto.');
    if (typeof options.isSlashGame !== 'boolean') throw new TypeError('INVALID_COMMAND_TYPE: La opción isSlashGame debe ser un booleano.');

    if (!options.embed) options.embed = {};
    if (!options.embed.title) options.embed.title = '¿Quién es este Pokémon?';
    if (!options.embed.color) options.embed.color = '#5865F2';

    if (!options.timeoutTime) options.timeoutTime = 60000;
    if (!options.winMessage) options.winMessage = '¡Lo adivinaste! Era un {pokemon}.';
    if (!options.loseMessage) options.loseMessage = 'Mejor suerte la próxima vez. Era un {pokemon}.';
    if (!options.errMessage) options.errMessage = '¡No se pudo obtener los datos del Pokémon! Por favor, intenta de nuevo.';

    if (typeof options.embed !== 'object') throw new TypeError('INVALID_EMBED: La opción embed debe ser un objeto.');
    if (typeof options.embed.title !== 'string') throw new TypeError('INVALID_EMBED: El título del embed debe ser un string.');
    if (typeof options.embed.color !== 'string') throw new TypeError('INVALID_EMBED: El color del embed debe ser un string.');
    if (typeof options.timeoutTime !== 'number') throw new TypeError('INVALID_TIME: La opción de tiempo de espera debe ser un número.');
    if (typeof options.winMessage !== 'string') throw new TypeError('INVALID_MESSAGE: El mensaje de victoria debe ser un string.');
    if (typeof options.loseMessage !== 'string') throw new TypeError('INVALID_MESSAGE: El mensaje de derrota debe ser un string.');

    super();
    this.options = options;
    this.message = options.message;
    this.pokemon = {};
  }

  async sendMessage(content) {
    if (this.options.isSlashGame) return await this.message.editReply(content).catch(e => {});
    else return await this.message.channel.send(content).catch(e => {});
  }

  async startGame() {
    if (this.options.isSlashGame || !this.message.author) {
      if (!this.message.deferred) await this.message.deferReply().catch(e => {});
      this.message.author = this.message.user;
      this.options.isSlashGame = true;
    }

    const result = await fetch('https://api.gamecord.xyz/pokemon').then(res => res.json()).catch(e => { return {}; });
    if (!result.data) return this.sendMessage({ content: this.options.errMessage });

    this.pokemon = result.data;

    const embed = new EmbedBuilder()
      .setColor(this.options.embed.color)
      .setTitle(this.options.embed.title)
      .setImage('attachment://question-image.png')
      .addFields({ name: 'Tipos', value: this.pokemon.types.join(', ') || 'Sin datos', inline: true })
      .addFields({ name: 'Habilidades', value: this.pokemon.abilities.join(', ') || 'Sin datos', inline: true })
      .setAuthor({ name: this.message.author.tag, iconURL: this.message.author.displayAvatarURL({ dynamic: true }) });

    const attachment = new AttachmentBuilder(this.pokemon.questionImage, { name: 'question-image.png' });
    const msg = await this.sendMessage({ embeds: [embed], files: [attachment] });

    const filter = (m) => m.author.id === this.message.author.id;
    const collector = this.message.channel.createMessageCollector({ idle: this.options.timeoutTime, filter: filter });

    collector.on('collect', (m) => {
      collector.stop();
      return this.gameOver(msg, m.content?.toLowerCase() === this.pokemon.name.toLowerCase());
    });

    collector.on('end', (_, reason) => {
      if (reason === 'idle') return this.gameOver(msg, false);
    });
  }

  async gameOver(msg, result) {
    const GuessThePokemonGame = { player: this.message.author, pokemon: this.pokemon };
    this.emit('gameOver', { result: result ? 'win' : 'lose', ...GuessThePokemonGame });

    if (!result) return msg.edit({ content: this.options.loseMessage.replace('{pokemon}', this.pokemon.name), embeds: [], attachments: [] });

    const embed = new EmbedBuilder()
      .setColor(this.options.embed.color)
      .setTitle(this.options.embed.title)
      .setImage('attachment://answer-image.png')
      .addFields({ name: 'Tipos', value: this.pokemon.types.join(', ') || 'Sin datos', inline: true })
      .addFields({ name: 'Habilidades', value: this.pokemon.abilities.join(', ') || 'Sin datos', inline: true })
      .setAuthor({ name: this.message.author.tag, iconURL: this.message.author.displayAvatarURL({ dynamic: true }) });

    const attachment = new AttachmentBuilder(this.pokemon.answerImage, { name: 'answer-image.png' });
    return msg.edit({ content: this.options.winMessage.replace('{pokemon}', this.pokemon.name), embeds: [embed], files: [attachment] });
  }
}
